src/types/methods.ts

```diff
@@ -0,0 +1,169 @@
+import type {
+  BuyItemByTypeAndroid,
+  BuyProductIos,
+  UserDataAmazon,
+} from '../modules';
+
+/** Sku is a string that uniquely identifies a product/subscription */
+export type Sku = string;
+
+export enum ProductType {
+  /** Subscription */
+  'subs' = 'subs',
+
+  /** Subscription */
+  'sub' = 'sub',
+
+  /** Consumable */
+  'inapp' = 'inapp',
+
+  /** Consumable */
+  'iap' = 'iap',
+}
+
+enum PaymentMethodIOS {
+  'FREETRIAL' = 'FREETRIAL',
+  'PAYASYOUGO' = 'PAYASYOUGO',
+  'PAYUPFRONT' = 'PAYUPFRONT',
+}
+
+enum PeriodUnitIOS {
+  'DAY' = 'DAY',
+  'WEEK' = 'WEEK',
+  'MONTH' = 'MONTH',
+  'YEAR' = 'YEAR',
+}
+
+export enum PurchaseStateAndroid {
+  UNSPECIFIED_STATE = 0,
+  PURCHASED = 1,
+  PENDING = 2,
+}
+
+/**
+ * Common interface for all modules.
+ */
+export interface NativeModuleProps {
+  /** Required method to start a payment provider connection */
+  initConnection(): Promise<boolean>;
+
+  /** Required method to end the payment provider connection */
+  endConnection(): Promise<boolean>;
+
+  /** addListener for NativeEventEmitter */
+  addListener(eventType: string): void;
+
+  /** removeListeners for NativeEventEmitter */
+  removeListeners(count: number): void;
+}
+
+/**
+ * Product or subscription item payload.
+ */
+export interface Discount {
+  identifier: string;
+  type: string;
+  numberOfPeriods: string;
+  price: string;
+  localizedPrice: string;
+  paymentMode: PaymentMethodIOS | '';
+  subscriptionPeriod: string;
+}
+
+export interface ProductCommon {
+  type: ProductType;
+  productId: string;
+  title: string;
+  description: string;
+  price: string;
+  currency: string;
+  localizedPrice: string;
+  countryCode?: string;
+}
+
+export interface ProductProduct extends ProductCommon {
```

Comment:
Why `ProductProduct` what was the problem with `Product`?

Reply:
Ah well! A big dilemma here.

We had [this type](https://github.com/dooboolab/react-native-iap/pull/1797/files/beca8d361c6aedd4a19efc21dbecd4ceffe2a198#diff-9463dd5d6cf72f3484ed3a3022651f3e8f1c0df3b03b553237a03c383bc867ccR169) before (`type Purchase = ProductPurchase | SubscriptionPurchase`). Instead of typing all the methods with `result: ProductPurchase | SubscriptionPurchase` we can then shortcut to `result: Purchase`. However, this is for a purchase response.

But for getting items (products or subscriptions) I had no idea on how to name the union type. Because we have products (consumable, non consumable) and subscriptions. which end up being `type Product = ProductProduct | SubscriptionProduct;` To keep consistency I wanted to find a better name, but couldn't think of anything else that would make sense